'set -b' should notify job state changes immediately

JohnoKing opened the pull request:
`set -b` currently does not notify job state changes when a background job finishes running. Reproducer and documentation for `set -b`:
```sh
$ set -b
$ sleep .1 &
# Notification doesn't appear when `sleep .1` finishes
```
```
  -b              The shell writes a message to standard error as soon it detects that
                  a background job completes rather than waiting until the next prompt.
```

This option has been broken since ksh93t 2008-07-25. Here is the change in that version that caused `set -b` to break:
```diff
        oerrno = errno;
        while(1)
        {
-               if(!(flags&WNOHANG) && !sh.intrap && waitevent && job.pwlist)
+               if(!(flags&WNOHANG) && !sh.intrap && job.pwlist)
                {
-                       if((*waitevent)(-1,-1L,0))
+                       sh_onstate(SH_TTYWAIT);
+                       if(waitevent && (*waitevent)(-1,-1L,0))
                                flags |= WNOHANG;
                }
                pid = waitpid((pid_t)-1,&wstat,flags);
+               sh_offstate(SH_TTYWAIT);
 
                /*
                 * some systems (linux 2.6) may return EINVAL
```

`SH_TTYWAIT` is always turned off even if it was previously on before the loop, which causes `set -b` to break since `SH_TTYWAIT` needs to be on for the shell to immediately show job state changes. The bugfix is from att/ast#1089, although one instance of `sh_offstate` has been placed behind `SHOPT_COSHELL` since it is only required if `SHOPT_COSHELL` is enabled.

JohnoKing committed:
'set -b' should notify job state changes immediately

src/cmd/ksh93/sh/jobs.c:
- Save the tty wait state and avoid changing it if TTYWAIT was already
  on to avoid breaking 'set -b'. This fixes a regression that was
  introduced in ksh93t 2008-07-25. The bugfix is from https://github.com/att/ast/pull/1089.
  The last 'sh_offstate' is inside of an '#if' directive because it is
  only required when ksh is compiled with SHOPT_COSHELL enabled.

src/cmd/ksh93/tests/pty.sh:
- Add a regression test for 'set -b' in interactive shells.

McDutchie commented on src/cmd/ksh93/tests/pty.sh:
```diff
@@ -549,5 +549,17 @@ r ^:test-2: true string\\r\\n$
+
+p :test-1:
+w set -b; sleep .01 &
+s 20
```
As `u` is documented to "read lines from the master until one matches *re*", I think you shouldn't need the `s` (sleep) here; it should keep reading/waiting until reaching either a match or EOF.

McDutchie commented on src/cmd/ksh93/sh/jobs.c:
```diff
@@ -374,11 +374,13 @@ int job_reap(register int sig)
 		flags = WUNTRACED|wcontinued;
 	shp->gd->waitevent = 0;
 	oerrno = errno;
+	int was_ttywait_on = sh_isstate(SH_TTYWAIT); /* save tty wait state */
```
This codebase usually declares all the variables used at the top of the function body; I'd like to stay consistent with that.

JohnoKing committed:
Apply requested changes

src/cmd/ksh93/sh/jobs.c:
- Move initial declaration of was_ttywait_on to start of job_reap.

src/cmd/ksh93/tests/pty.sh:
- Remove 's 20' since 'u' should keep reading until EOF.

McDutchie commented on src/cmd/ksh93/sh/jobs.c:
```diff
@@ -348,6 +348,7 @@ int job_reap(register int sig)
 	Cojob_t		*cjp;
 	int		cojobs;
 	long		cotimeout = sig?0:-1;
+	int was_ttywait_on = sh_isstate(SH_TTYWAIT); /* save tty wait state */
```
This moves it within `#if SHOPT_COSHELL`...`#endif`, which I don't think is right.

I was thinking more along the lines of  a`int was_ttywait_on;` declaration on line 347, then actually save the state on line 377... that is the practice in most of the rest of the code.

JohnoKing committed:
Move was_ttywait_on out of SHOPT_COSHELL
